### PR TITLE
Bxmsdoc 2131 Business Optimizer Planning Problem

### PR DESCRIPTION
**Pull request by gsheldon**
Slightly edited and modularized for user story.

Same content as in BRP Guide.

### CODE REVIEW

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+* Maximized profits - the optimal goal results in the highest possible profit.
+* Minimized ecological footprint - the optimal goal has the least amount of environmental impact.
+* Maximized satisfaction for employees or customers - the optimal goal prioritizes the needs of employees or customers.
+
+The ability to achieve these goals relies on the number of resources available, such as:
+
+* The number of people
+* Amount of time
+* Budget
+* Physical assets, for example, machinery, vehicles, computers, buildings, etc
```
Comment (emmurphy1) [inline docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc]: IBM SG: etc. n
Do not use. Use “and so on”

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+Specific constraints related to these resources must also be taken into account, such as the number of hours a person works, their ability to use certain machines, or compatibility between pieces of equipment.
```
Comment (emmurphy1) [inline docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc]: must also be.. passive
Under the hood - slang?

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+There are several categories of solutions:
+
+Possible solution::
+A possible solution is any solution, whether or not it breaks any number of constraints. Planning problems tend to have an incredibly large number of possible solutions. Many of those solutions are worthless.
+
+Feasible solution::
+A feasible solution is a solution that does not break any (negative) hard constraints. The number of feasible solutions tends to be relative to the number of possible solutions. Sometimes there are no feasible solutions. Every feasible solution is a possible solution.
+
+Optimal solution::
+An optimal solution is a solution with the highest score. Planning problems tend to have 1 or a few optimal solutions. There is always at least 1 optimal solution, even in the case that there are no feasible solutions and the optimal solution is not feasible.
```
Comment (emmurphy1) [inline docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc]: tend to - casual. Consider 'usually' or 'often'. For numbers under 10, spell out (one). Consider' Planning problems usually have a few options solutions. They always have at least one..

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+Best solution found::
+The best solution found is the solution with the highest score found by an implementation in a given amount of time. The best solution found is likely to be feasible and, given enough time, it's an optimal solution.
+
+Counterintuitively, the number of possible solutions is huge (if calculated correctly), even with a small data set.
+
+In the examples provided in the `planner-engine` distribution folder, most instances have a significantly higher number of possible solutions than the minimal number of atoms in the known universe (10^80). As there is no guaranteed way to find the optimal solution, any implementation is forced to evaluate at least a subset of all those possible solutions.
```
Comment (emmurphy1) [inline docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc]: I'll leave it up to you, but think this is too chatting for tech doc:
minimal number of atoms in the known universe (10^80). Simplly , most instances have a huge, enormous, vast...
all of those possible.

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-NP-complete-vs-NP-hard-con.adoc [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+But if they find one for 1 NP-complete problem, it will work for every NP-complete problem.
+
+In fact, there's a $ 1,000,000 reward for anyone that proves if http://en.wikipedia.org/wiki/P_%3D_NP_problem[such a silver bullet actually exists or not].
+
+====
+////
+
+The implication is that solving your problem is probably harder than you anticipated, because the two common techniques will not suffice:
+
+* A Brute Force algorithm (even a more advanced variant) will take too long.
```
Comment (emmurphy1) [inline docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-NP-complete-vs-NP-hard-con.adoc]: Why cap Brute Force?

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-hard-vs-soft-constraints-con.adoc [hunk 6]
```diff
@@ -0,0 +1,26 @@
+[id='optimizer-hard-vs-soft-constraints-con']
+= A planning problem has (hard and soft) constraints
```
Comment (emmurphy1) [inline docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-hard-vs-soft-constraints-con.adoc]: Why (hard and soft)? Why not ...has hard and soft contraints?

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-hard-vs-soft-constraints-con.adoc [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+Usually, a planning problem has at least two levels of constraints:
+
+* A _(negative) hard constraint_ must not be broken. 
++
+For example, one teacher can not teach two different lessons at the same time.
+* A _(negative) soft constraint_ should not be broken if it can be avoided. 
++
+For example, Teacher A does not like to teach on Friday afternoon.
```
Comment (emmurphy1) [inline docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-hard-vs-soft-constraints-con.adoc]: afternoons

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-hard-vs-soft-constraints-con.adoc [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+Some problems also have positive constraints:
+
+* A _positive soft constraint (or reward)_ should be fulfilled if possible. 
++
+For example, Teacher B likes to teach on Monday morning.
```
Comment (emmurphy1) [inline docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-hard-vs-soft-constraints-con.adoc]: mornings

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-hard-vs-soft-constraints-con.adoc [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+Some basic problems only have hard constraints.
+Some problems have three or more levels of constraints, for example, hard, medium, and soft constraints.
+
+These constraints define the _score calculation_ (otherwise known as the __fitness function__) of a planning problem.
+Each solution of a planning problem can be graded with a score. With the Business Optimizer, score constraints are written in an Object Oriented language, such as Java code or Drools rules.
```
Comment (emmurphy1) [inline docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-hard-vs-soft-constraints-con.adoc]: object oriented l/c

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-hard-vs-soft-constraints-con.adoc [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+Such code is flexible and scalable.
```
Comment (emmurphy1) [inline docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-hard-vs-soft-constraints-con.adoc]: This type of code...

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-NP-complete-vs-NP-hard-con.adoc [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+* A quick algorithm, for example in bin packing, _putting in the largest items first_ will return a solution that is far from optimal.
```
Comment (emmurphy1) [inline docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-NP-complete-vs-NP-hard-con.adoc]: This is a little confusing. I don't understand putting the largest items first.

File: docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc [hunk 12]
```diff
@@ -1,10 +1,11 @@
 [id='assembly_business_optimizer']
```
Comment (sterobin) [inline docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc]: Should be [id='assembly_business-optimizer']  i.e., the "assembly_" prefix is the only thing in our anchor names that has the underscore. Everything else is dashes only.

File: docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc [hunk 13]
```diff
@@ -1,10 +1,11 @@
...
 
+
 include::product-shared-docs/document-attributes.adoc[]
 
 = Solving planning problems using the Business Optimizer
 
-:imagesdir: topics/product-shared-docs/images
+:imagesdir: topics/product-shared-docs/images/
 :context: business-optimizer
```
Comment (sterobin) [inline docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc]: Missing the author-group include here, immediately following :context: line. So should be:
...
:context: business-optimizer
include::product-shared-docs/author-group.adoc[]

....

File: docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc [hunk 14]
```diff
@@ -16,8 +17,11 @@ Use this document to set up and configure {PLANNER} functionality.
 The *Employee Rostering* sample project has been created in {CENTRAL} to demonstrate Business Optimizer capabilities. To view this project, you must have {PRODUCT} installed. For more information about setting up and installing {PRODUCT} and {CENTRAL}, see _Getting started with decision services_.
 
 // Modules
-
+//About
 include::product-business-resource-planner-guide/optimizer-about-business-optimizer-con.adoc[leveloffset=+1]
```
Comment (sterobin) [inline docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc]: Just a note for this actual content file (optimizer-about...) since it's not part of the PR:
A couple of the listed items use colons, a couple periods. Should be colons all around (e.g., "Employee/Patient Rosters:") I noticed it in the output.

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+* Budget
+* Physical assets, for example, machinery, vehicles, computers, buildings, etc
+
+Specific constraints related to these resources must also be taken into account, such as the number of hours a person works, their ability to use certain machines, or compatibility between pieces of equipment. 
+ 
+The Business Optimizer helps Java programmers solve constraint satisfaction problems efficiently. Under the hood, it combines optimization heuristics and metaheuristics with efficient score calculation.
+
+
+[id='optimizer-planning-problem-search-space-con']
+== A Planning Problem Has a Huge Search Space
```
Comment (sterobin) [inline docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc]: This section should technically be a separate module that you include in the assembly. Each module should almost always have only one heading (meaning one genuine = Heading).

Also, the title should be sentence case, as you've correctly done everywhere else (A planning problem has a huge search space).

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+In the examples provided in the `planner-engine` distribution folder, most instances have a significantly higher number of possible solutions than the minimal number of atoms in the known universe (10^80). As there is no guaranteed way to find the optimal solution, any implementation is forced to evaluate at least a subset of all those possible solutions.
+
+The Business Optimizer supports several optimization algorithms to efficiently wade through that incredibly large number of possible solutions.
+
+Depending on the use case, some optimization algorithms perform better than others, but it is impossible to tell in advance. Using the Business Optimizer, you can switch the optimization algorithm by changing the solver configuration in a few lines of XML or code.
+
+
+include::optimizer-NP-complete-vs-NP-hard-con.adoc[leveloffset=+1]
+
+include::optimizer-hard-vs-soft-constraints-con.adoc[leveloffset=+1]
```
Comment (sterobin) [inline docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc]: So if these two includes must be part of the "A planning problem has a huge search space" module for it to be complete (i.e., it truly can't and won't ever be able to be reused without these two other dependent modules included in it) then this is fine. But if that's not the case, then move these includes to the assembly, following this module's include (which again should be a separate include, rather than nested in "What is a planning problem?". Probably seems tedious, but it's all about creating building blocks to be as reusable as possible. As you know.

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc [hunk 17]
```diff
@@ -0,0 +1,53 @@
+[id='optimizer-planning-problem-con']
+= What is a planning problem?
+
+
+A planning problem has an optimal goal, based on limited resources and under specific constraints. Optimal goals can be any number of things, such as:
```
Comment (sterobin) [inline docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc:5]: I'll only mention it once since it comes up a few other places, and since I think Emily already mentioned it. IBM SG is big on complete sentences before colons, without ending with "such as", "the following", etc. Can be resolved as easily as "such as the following examples:" I personally am not a huge fan and actually prefer what you've done, as an economy. But just sayin'. I've confirmed my stuff accordingly as well.

File: docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc [hunk 18]
```diff
@@ -16,8 +17,11 @@ Use this document to set up and configure {PLANNER} functionality.
...
+include::product-business-resource-planner-guide/optimizer-planning-problem-con.adoc[leveloffset=+1]
+
+//Downloading and installing
 include::product-business-resource-planner-guide/optimizer-download-proc.adoc[leveloffset=+1]
```
Comment (sterobin) [inline docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc]: Again, since the content file is not included, just a note that I _think_ we're going with the passive on prerequisites. Makes me crazy and I much prefer the active, like you have it (Download and install ...), but passive is the decided style I think for prereqs because it leans toward the thing already being done, and not always by the other in some cases. If it's _not_ likely to already have been done (which would call for active), then the argument is that it shouldn't be a prereq, but part of the procedure. That's the logic for the decision anyway. So the prereqs here, e.g., would be "{PRODUCT} is (or has been) downloaded and installed from..." Emily and I were just talking about this the other day. The choice is yours, but just a heads-up on the likely direction we're going. Emily should be clarifying sometime down the road.

Only mentioning this here, btw, just as a general note.

File: docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc [hunk 19]
```diff
@@ -16,8 +17,11 @@ Use this document to set up and configure {PLANNER} functionality.
...
 include::product-business-resource-planner-guide/optimizer-engine-installation-proc.adoc[leveloffset=+1]
```
Comment (sterobin) [inline docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc]: And again since this file isn't part of the PR, just mentioning that the context statement following the header should provide more of a statement of the why and wherefore, i.e., it's context, rather than a command. Like you've done well elsewhere. And shouldn't end in a colon.

Something as simple as (just my whack at it):
= Installing the {PLANNER} engine in your application
You can install {PLANNER} with Maven and with various other applications, such as Gradle, Ivy, or Buildr, to optimize your existing planning environment.

Or something. I would assess all context statements, especially in procedures (the biggest pitfall) to make sure they are genuine context statements. Here's the [templates page](https://mojo.redhat.com/docs/DOC-1099486#jive_content_id_Master_source_files_in_asciidoc_format) too, if you need it.

Also, just a note that the prereq for this section ("Get the {PLANNER} JARs ...") is an example of a prereq that could arguably just be the first step in the actual procedure. The argument, again, for passive prereqs is that if you can't make it passive and sound normal (where active is really the only way) then it's highly likely it should be part of the procedure. I don't know for sure in this case. Just sayin'. Your call.

Lastly, the "Installing with Ant" sub-heading should be it's only module, included in the assembly. And the context statement should likewise be updated like I've described above.

**Review by sterobin — COMMENTED** [review]
@gsheldon, these comments are what I have for now. Feel free to shoot it back to me later when the docs more formulated. I'm happy to help.

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-NP-complete-vs-NP-hard-con.adoc (hunk 5 above)
Comment (gsheldon) [reply docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-NP-complete-vs-NP-hard-con.adoc]: ah! Missed that one, ta.

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-NP-complete-vs-NP-hard-con.adoc (hunk 11 above)
Comment (gsheldon) [reply docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-NP-complete-vs-NP-hard-con.adoc]: I'm including an external link to the Wikipedia entry for the Bin packing problem, as it's too complicated to explain in our doc and not really necessary. This should give it some context at least.

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-hard-vs-soft-constraints-con.adoc (hunk 6 above)
Comment (gsheldon) [reply docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-hard-vs-soft-constraints-con.adoc]: I guess because sometimes they can have only hard constraints, or sometimes hard, medium, and soft constraints. It's probably better as just constraints, so I'll change the title to that.

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc (hunk 3 above)
Comment (gsheldon) [reply docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc]: oh man, I thought I got rid of all of the numbers, but clearly a few slipped through!

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc (hunk 4 above)
Comment (gsheldon) [reply docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc]: hmm, yeah it's impressive but it probably works better in the community content than here.

File: docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc (hunk 12 above)
Comment (gsheldon) [reply docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc]: oops, missed that!

File: docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc (hunk 18 above)
Comment (gsheldon) [reply docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc]: Interesting. I tend to not bother getting involved in these debates because they take up too much time and I can't engage in real time with everyone else, so I'll change it here anyway since it sounds like you and Emily are already doing it.

File: docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc (hunk 19 above)
Comment (gsheldon) [reply docs/product-assembly_business-optimizer/src/main/asciidoc/main.adoc]: I'm reluctant to break Ant into it's own module as it's a very niche use case. This will *never* be reused anywhere else. I'd prefer to maybe reformat into some kind of alternative step, as it's still installing the engine, it's just using a different application than 99% of users would use.

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc (hunk 15 above)
Comment (gsheldon) [reply docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc]: Forgot to change that one, thanks,

File: docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc (hunk 17 above)
Comment (gsheldon) [reply docs/product-business-resource-planner-guide/src/main/asciidoc/optimizer-planning-problem-con.adoc:5]: Fixed in a couple of places, some other parts are still a WIP
